docs/content/concepts/io-management/io-managers.mdx

```diff
@@ -310,31 +321,59 @@ def a_job():
 
 ```python file=/concepts/io_management/custom_io_manager.py startafter=start_io_manager_marker endbefore=end_io_manager_marker
-from dagster import IOManager, io_manager
+from dagster import ConfigurableIOManager
 
 
-class MyIOManager(IOManager):
+class MyIOManager(ConfigurableIOManager):
+    path_prefix: List[str] = []
+
     def _get_path(self, context) -> str:
-        return "/".join(context.asset_key.path)
+        return "/".join(self.path_prefix + context.asset_key.path)
 
     def handle_output(self, context, obj):
         write_csv(self._get_path(context), obj)
 
     def load_input(self, context):
         return read_csv(self._get_path(context))
+```
 
+The provided `context` argument for `handle_output` is an <PyObject module="dagster" object="OutputContext" />. The provided `context` argument for `load_input` is an <PyObject module="dagster" object="InputContext" />. The linked API documentation lists all the fields that are available on these objects.
 
-@io_manager
-def my_io_manager():
-    return MyIOManager()
-```
+### Using an IO manager factory
 
-The <PyObject module="dagster" object="io_manager" /> decorator behaves nearly identically to the <PyObject module="dagster" object="resource" /> decorator. It yields an <PyObject module="dagster" object="IOManagerDefinition" />, which is a subclass of <PyObject module="dagster" object="ResourceDefinition" /> that will produce an <PyObject module="dagster" object="IOManager" />.
+If your IO manager is more complex, or needs to manage internal state, it may make sense to split out the IO manager definition from its configuration. In this case, you can use `ConfigurableIOManagerFactory`, which specifies config schema and implements a factory function that takes the config and returns an IO manager.
 
-The provided `context` argument for `handle_output` is an <PyObject module="dagster" object="OutputContext" />. The provided `context` argument for `load_input` is an <PyObject module="dagster" object="InputContext" />. The linked API documentation lists all the fields that are available on these objects.
+```python file=/concepts/io_management/custom_io_manager.py startafter=start_io_manager_factory_marker endbefore=end_io_manager_factory_marker
+from dagster import IOManager, ConfigurableIOManagerFactory
+import requests
+
+
+class ExternalIOManager(IOManager):
+    def __init__(self, api_token):
+        self._api_token = api_token
+
+    def handle_output(self, context, obj):
+        ...
+
+    def load_input(self, context):
+        ...
+
+
+class ConfigurableExternalIOManager(ConfigurableIOManagerFactory):
+    username: str
+    password: str
+
+    def create_io_manager(self, context) -> ExternalIOManager:
+        api_token = requests.get(
+            "https://my-api.com/token", auth=(self.username, self.password)
+        ).json()
+        return ExternalIOManager(api_token)
+```
```

Comment:
i found this example a bit abstract and hard to grasp why i'd bother with the factor pattern. maybe worth switching to a more realistic io manager example, or keep it but also link to concrete examples like "Custom filesystem-based IO Manager" down below.

Reply:
Could be as simple as adding some sort of cache as private variable to the `ExternalIOManager` and note that one could not add that to a `ConfigurableIOManager` without erroring.

Reply:
Updated example to use a cache.